api/src/main/java/org/openmrs/api/impl/TagServiceImpl.java

```diff
@@ -0,0 +1,158 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.impl;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.OpenmrsObject;
+import org.openmrs.api.UserService;
+import org.openmrs.Tag;
+import org.openmrs.api.TagService;
+import org.openmrs.api.db.TagDAO;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.ArrayList;
+
+@Transactional(readOnly = true)
+public class TagServiceImpl extends BaseOpenmrsService implements TagService {
+	
+	protected final Log log = LogFactory.getLog(this.getClass());
+	
+	private TagDAO dao;
+	
+	UserService userService;
+	
+	/**
+	 * Injected in moduleApplicationContext.xml
+	 */
+	public void setDao(TagDAO dao) {
+		this.dao = dao;
+	}
+	
+	/**
+	 * Injected in moduleApplicationContext.xml
+	 */
+	public void setUserService(UserService userService) {
+		this.userService = userService;
+	}
+	
+	@Override
+	@Transactional
+	public Tag getTagByUuid(String uuid) {
+		return dao.getTagByUuid(uuid);
+	}
+	
+	@Override
+	public Tag saveTag(Tag tag) {
+		return dao.saveTag(tag);
+	}
+	
+	@Override
+	public void removeTag(Tag tag) {
+		dao.deleteTag(tag);
+	}
+	
+	@Override
+	@Transactional
+	public List<Tag> getAllTags() {
+		return dao.getAllTags();
+	}
+	
+	@Override
+	@Transactional
+	public Tag getTagById(int id) {
+		return dao.getTagById(id);
+	}
+	
+	@Override
+	@Transactional
+	public List<Tag> getTags(String tag) {
+		return dao.getTags(tag);
+	}
+	
+	@Override
+	public boolean object_exits(String object_uuid, String object_type) throws ClassNotFoundException {
+		Object object = dao.object_exists(object_uuid, object_type);
+		if (!object.equals(null))
+			return true;
+		return false;
+	}
+	
+	@Override
+	public Tag addTag(OpenmrsObject openmrsObject, String tag) throws Exception {
+		if (duplicateTag(openmrsObject, tag)) {
+			log.warn("duplicate Tag for " + openmrsObject);
+			return null;
+		} else {
+			Tag tag1 = new Tag(tag, openmrsObject.getUuid(), openmrsObject.getClass().toString().substring(6));
+			return saveTag(tag1);
+		}
+	}
+	
+	/**
+	 * Validation Method which logs a warning if u try to add a duplicate tag to an OpenmrsObject
+	 */
+	public boolean duplicateTag(OpenmrsObject openmrsObject, String tag) throws Exception {
+		List<String> list = getTags(openmrsObject);
```

Comment:
You should be calling getTags(List,List,true) with the object type and uuid as the only elements in the passed in lists and only have to check that the returned list is empty so you don't have to loop over like this.